File: Problems/Leetcode/3001-4000/3301-3400/3366.cpp
#include <bits/stdc++.h>

#include "debug.h"

using namespace std;

static constexpr int MOD = 1'000'000'007;

class Solution {
public:
    int minArraySum(vector<int>& nums, int k, int op1, int op2) {
        vector f(op1 + 1, vector<int>(op2 + 1));
        for (int x : nums) {
            for (int p = op1; p >= 0; p--) {
                for (int q = op2; q >= 0; q--) {
                    int res = f[p][q] + x;
                    if (p) {
                        res = min(res, f[p - 1][q] + (x + 1) / 2);
                    }
                    if (q && x >= k) {
                        res = min(res, f[p][q - 1] + x - k);
                        if (p) {
                            int y = (x - k + 1) / 2;
                            if ((x + 1) / 2 >= k) {
                                y = min(y, (x + 1) / 2 - k);
                            }
                            res = min(res, f[p - 1][q - 1] + y);
                        }
                    }
                    f[p][q] = res;
                }
            }
        }
        return f[op1][op2];
    }
};

int main() {
    Solution sol;

    return 0;
}

File: Problems/Leetcode/3001-4000/3301-3400/3366.py
from functools import cache
from typing import List

fmin = lambda x, y: x if x < y else y


class Solution:
    def minArraySum(self, nums: List[int], k: int, op1: int, op2: int) -> int:
        s = sum(nums)
        n = len(nums)

        @cache
        def dfs(i: int, op1: int, op2: int):
            if i == n:
                return 0
            if op1 < 0 or op2 < 0:
                return s

            res = dfs(i + 1, op1, op2) + nums[i]
            if op1 > 0:
                res = fmin(res, dfs(i + 1, op1 - 1, op2) + (nums[i] + 1) // 2)
            if op2 > 0 and nums[i] >= k:
                res = fmin(res, dfs(i + 1, op1, op2 - 1) + nums[i] - k)
            if op1 > 0 and op2 > 0 and (nums[i] + 1) // 2 >= k:
                res = fmin(res, dfs(i + 1, op1 - 1, op2 - 1) + (nums[i] + 1) // 2 - k)
            if op2 > 0 and nums[i] >= k and op1 > 0:
                res = fmin(res, dfs(i + 1, op1 - 1, op2 - 1) + (nums[i] - k + 1) // 2)
            return res

        ans = dfs(0, op1, op2)
        dfs.cache_clear()
        return ans
